File: apps/app-frontend/src/helpers/logs.js
/**
 * All theseus API calls return serialized values (both return values and errors);
 * So, for example, addDefaultInstance creates a blank instance object, where the Rust struct is serialized,
 *  and deserialized into a usable JS object.
 */
import { invoke } from '@tauri-apps/api/core'
import { save } from '@tauri-apps/plugin-dialog'

/*
A log is a struct containing the filename string and optional output, as follows:

pub struct Logs {
    pub filename:  String,
    pub output: Option<String>,
}
*/

/// Get all logs that exist for a given instance
/// This is returned as an array of Log objects, sorted by filename (the folder name, when the log was created)
export async function get_logs(instanceId, clearContents) {
	return await invoke('plugin:logs|logs_get_logs', { instanceId, clearContents })
}

/// Get an instance's log by filename
export async function get_logs_by_filename(instanceId, logType, filename) {
	return await invoke('plugin:logs|logs_get_logs_by_filename', { instanceId, logType, filename })
}

/// Get an instance's log text only by filename
export async function get_output_by_filename(instanceId, logType, filename) {
	return await invoke('plugin:logs|logs_get_output_by_filename', {
		instanceId,
		logType,
		filename,
	})
}

/// Delete an instance's log by filename
export async function delete_logs_by_filename(instanceId, logType, filename) {
	return await invoke('plugin:logs|logs_delete_logs_by_filename', {
		instanceId,
		logType,
		filename,
	})
}

/// Delete all logs for a given instance
export async function delete_logs(instanceId) {
	return await invoke('plugin:logs|logs_delete_logs', { instanceId })
}

/// Get the latest log for a given instance and cursor (startpoint to read within the file)
/// Returns:
/*
  {
    cursor: u64
    output: String
    new_file: bool <- the cursor was too far, meaning that the file was likely rotated/reset. This signals to the frontend to clear the log and start over with this struct.
  }
*/

// From the launcher's launcher_log.txt directly
export async function get_latest_log_cursor(instanceId, cursor) {
	return await invoke('plugin:logs|logs_get_latest_log_cursor', { instanceId, cursor })
}

/// Read Minecraft's logs/latest.log from a cursor.
export async function get_minecraft_latest_log_cursor(instanceId, cursor) {
	return await invoke('plugin:logs|logs_get_minecraft_latest_log_cursor', { instanceId, cursor })
}

/// Get all buffered live log lines for an instance from the Rust ring buffer
export async function get_live_log_buffer(instanceId) {
	return await invoke('plugin:logs|logs_get_live_log_buffer', { instanceId })
}

/// Clear the live log buffer for an instance on the Rust side
export async function clear_log_buffer(instanceId) {
	return await invoke('plugin:logs|logs_clear_live_log_buffer', { instanceId })
}

/// Collect and locally analyze the logs from an instance's latest run.
export async function analyze_crash(instanceId) {
	return await invoke('plugin:logs|logs_analyze_crash', { instanceId })
}

export async function get_crash_analysis_ai_settings() {
	return await invoke('plugin:logs|logs_get_crash_analysis_ai_settings')
}

export async function update_crash_analysis_ai_settings(settings) {
	return await invoke('plugin:logs|logs_update_crash_analysis_ai_settings', { settings })
}

export async function explain_crash_with_ai(instanceId) {
	return await invoke('plugin:logs|logs_explain_crash_with_ai', { instanceId })
}

export async function undo_added_mod(instanceId, filename, expectedHash) {
	return await invoke('plugin:logs|logs_undo_added_mod', { instanceId, filename, expectedHash })
}

export async function get_log_share_settings() {
	return await invoke('plugin:logs|logs_get_log_share_settings')
}

export async function update_log_share_settings(settings) {
	return await invoke('plugin:logs|logs_update_log_share_settings', { settings })
}

export async function logshare_upload_crash(instanceId) {
	return await invoke('plugin:logs|logs_logshare_upload_crash', { instanceId })
}

export async function logshare_get_insights(id) {
	return await invoke('plugin:logs|logs_logshare_get_insights', { id })
}

export async function logshare_analyse_crash_direct(instanceId) {
	return await invoke('plugin:logs|logs_logshare_analyse_crash_direct', { instanceId })
}

export async function logshare_ai_analyze_stored(instanceId, id) {
	return await invoke('plugin:logs|logs_logshare_ai_analyze_stored', { instanceId, id })
}

export async function logshare_ai_analyze_direct(instanceId) {
	return await invoke('plugin:logs|logs_logshare_ai_analyze_direct', { instanceId })
}

export async function logshare_delete(id, token) {
	return await invoke('plugin:logs|logs_logshare_delete', { id, token })
}

export async function list_shared_logs() {
	return await invoke('plugin:logs|logs_list_shared_logs')
}

export async function record_shared_log(log) {
	return await invoke('plugin:logs|logs_record_shared_log', { log })
}

export async function delete_shared_log(id, token) {
	return await invoke('plugin:logs|logs_delete_shared_log', { id, token })
}

/// Export the censored files and local analysis from an instance's latest run.
export async function export_crash_context(instanceId, instanceName) {
	const timestamp = new Date().toISOString().replace(/[:.]/g, '-')
	const outputPath = await save({
		defaultPath: `${instanceName || 'Minecraft'} crash context ${timestamp}.zip`,
		filters: [{ name: 'ZIP archive', extensions: ['zip'] }],
	})
	if (!outputPath) return false
	await invoke('plugin:logs|logs_export_crash_context', { instanceId, outputPath })
	return true
}

File: apps/app/src/api/logs.rs
use crate::api::Result;
use async_zip::tokio::write::ZipFileWriter;
use std::path::PathBuf;
use theseus::logs::LogType;
use theseus::logs::{
    self, CensoredString, CrashAnalysis, CrashAnalysisAiExplanation,
    CrashAnalysisAiSettings, LatestLogCursor, LogShareSettings,
    LogShareUploadResponse, Logs, SharedLog,
};

/*
A log is a struct containing the filename string, stdout, and stderr, as follows:

pub struct Logs {
    pub filename:  String,
    pub stdout: String,
    pub stderr: String,
}
*/

pub fn init<R: tauri::Runtime>() -> tauri::plugin::TauriPlugin<R> {
    tauri::plugin::Builder::new("logs")
        .invoke_handler(tauri::generate_handler![
            logs_get_logs,
            logs_get_logs_by_filename,
            logs_get_output_by_filename,
            logs_delete_logs,
            logs_delete_logs_by_filename,
            logs_get_latest_log_cursor,
            logs_get_minecraft_latest_log_cursor,
            logs_get_live_log_buffer,
            logs_clear_live_log_buffer,
            logs_analyze_crash,
            logs_get_crash_analysis_ai_settings,
            logs_update_crash_analysis_ai_settings,
            logs_explain_crash_with_ai,
            logs_undo_added_mod,
            logs_export_crash_context,
            logs_get_log_share_settings,
            logs_update_log_share_settings,
            logs_logshare_upload_crash,
            logs_logshare_get_insights,
            logs_logshare_analyse_crash_direct,
            logs_logshare_ai_analyze_stored,
            logs_logshare_ai_analyze_direct,
            logs_logshare_delete,
            logs_list_shared_logs,
            logs_record_shared_log,
            logs_delete_shared_log,
        ])
        .build()
}

/// Get all logs for an instance, sorted by filename.
#[tauri::command]
pub async fn logs_get_logs(
    instance_id: &str,
    clear_contents: Option<bool>,
) -> Result<Vec<Logs>> {
    let val = logs::get_logs(instance_id, clear_contents).await?;

    Ok(val)
}

/// Get a log struct for an instance by filename.
#[tauri::command]
pub async fn logs_get_logs_by_filename(
    instance_id: &str,
    log_type: LogType,
    filename: String,
) -> Result<Logs> {
    Ok(logs::get_logs_by_filename(instance_id, log_type, filename).await?)
}

/// Get the output for an instance by filename.
#[tauri::command]
pub async fn logs_get_output_by_filename(
    instance_id: &str,
    log_type: LogType,
    filename: String,
) -> Result<CensoredString> {
    Ok(logs::get_output_by_filename(instance_id, log_type, &filename).await?)
}

/// Delete all logs for an instance.
#[tauri::command]
pub async fn logs_delete_logs(instance_id: &str) -> Result<()> {
    Ok(logs::delete_logs(instance_id).await?)
}

/// Delete a log for an instance by filename.
#[tauri::command]
pub async fn logs_delete_logs_by_filename(
    instance_id: &str,
    log_type: LogType,
    filename: String,
) -> Result<()> {
    Ok(logs::delete_logs_by_filename(instance_id, log_type, &filename).await?)
}

/// Get live log from a cursor
#[tauri::command]
pub async fn logs_get_latest_log_cursor(
    instance_id: &str,
    cursor: u64, // 0 to start at beginning of file
) -> Result<LatestLogCursor> {
    Ok(logs::get_latest_log_cursor(instance_id, cursor).await?)
}

/// Get Minecraft's logs/latest.log from a cursor.
#[tauri::command]
pub async fn logs_get_minecraft_latest_log_cursor(
    instance_id: &str,
    cursor: u64, // 0 to start at beginning of file
) -> Result<LatestLogCursor> {
    Ok(
        logs::get_generic_live_log_cursor(instance_id, "latest.log", cursor)
            .await?,
    )
}

/// Get all buffered live log lines for an instance.
#[tauri::command]
pub async fn logs_get_live_log_buffer(
    instance_id: &str,
) -> Result<CensoredString> {
    Ok(logs::get_live_log_buffer(instance_id).await?)
}

/// Clear the live log buffer for an instance.
#[tauri::command]
pub async fn logs_clear_live_log_buffer(instance_id: &str) -> Result<()> {
    logs::clear_live_log_buffer(instance_id);
    Ok(())
}

/// Collect and locally analyze the files produced by the instance's latest run.
#[tauri::command]
pub async fn logs_analyze_crash(instance_id: &str) -> Result<CrashAnalysis> {
    Ok(logs::analyze_crash(instance_id).await?)
}

#[tauri::command]
pub async fn logs_get_crash_analysis_ai_settings()
-> Result<CrashAnalysisAiSettings> {
    Ok(logs::get_crash_analysis_ai_settings().await?)
}

#[tauri::command]
pub async fn logs_update_crash_analysis_ai_settings(
    settings: CrashAnalysisAiSettings,
) -> Result<()> {
    Ok(logs::update_crash_analysis_ai_settings(settings).await?)
}

#[tauri::command]
pub async fn logs_explain_crash_with_ai(
    instance_id: &str,
) -> Result<CrashAnalysisAiExplanation> {
    Ok(logs::explain_crash_with_ai(instance_id).await?)
}

#[tauri::command]
pub async fn logs_undo_added_mod(
    instance_id: &str,
    filename: &str,
    expected_hash: &str,
) -> Result<()> {
    Ok(logs::undo_added_mod(instance_id, filename, expected_hash).await?)
}

/// Export the latest run's censored diagnostic context as a ZIP archive.
#[tauri::command]
pub async fn logs_export_crash_context(
    instance_id: &str,
    output_path: PathBuf,
) -> Result<()> {
    let analysis = logs::analyze_crash(instance_id).await?;
    let archive = tokio::fs::File::create(&output_path).await?;
    let mut writer = ZipFileWriter::with_tokio(archive);
    let report = serde_json::to_vec_pretty(&analysis).map_err(|error| {
        theseus::Error::from(theseus::ErrorKind::OtherError(format!(
            "Failed to serialize crash analysis: {error}"
        )))
    })?;
    crate::api::utils::write_zip_entry(&mut writer, "analysis.json", &report)
        .await?;
    for source in &analysis.sources {
        let safe_name = source.filename.replace(['/', '\\'], "_");
        crate::api::utils::write_zip_entry(
            &mut writer,
            &format!("logs/{safe_name}"),
            source.content.as_str().as_bytes(),
        )
        .await?;
    }
    writer.close().await.map_err(crate::api::utils::zip_error)?;
    Ok(())
}

#[tauri::command]
pub async fn logs_get_log_share_settings() -> Result<LogShareSettings> {
    Ok(logs::get_log_share_settings().await?)
}

#[tauri::command]
pub async fn logs_update_log_share_settings(
    settings: LogShareSettings,
) -> Result<()> {
    Ok(logs::update_log_share_settings(settings).await?)
}

#[tauri::command]
pub async fn logs_logshare_upload_crash(
    instance_id: &str,
) -> Result<LogShareUploadResponse> {
    Ok(logs::upload_crash(instance_id).await?)
}

#[tauri::command]
pub async fn logs_logshare_get_insights(id: &str) -> Result<serde_json::Value> {
    Ok(logs::get_insights(id).await?)
}

#[tauri::command]
pub async fn logs_logshare_analyse_crash_direct(
    instance_id: &str,
) -> Result<serde_json::Value> {
    Ok(logs::analyse_crash_direct(instance_id).await?)
}

#[tauri::command]
pub async fn logs_logshare_ai_analyze_stored(
    instance_id: &str,
    id: &str,
) -> Result<String> {
    Ok(logs::ai_analyze_stored(instance_id, id).await?)
}

#[tauri::command]
pub async fn logs_logshare_ai_analyze_direct(
    instance_id: &str,
) -> Result<String> {
    Ok(logs::ai_analyze_direct(instance_id).await?)
}

#[tauri::command]
pub async fn logs_logshare_delete(id: &str, token: &str) -> Result<()> {
    logs::delete_log(id, token).await?;
    Ok(())
}

#[tauri::command]
pub async fn logs_list_shared_logs() -> Result<Vec<SharedLog>> {
    Ok(logs::list_shared_logs().await?)
}

#[tauri::command]
pub async fn logs_record_shared_log(log: SharedLog) -> Result<()> {
    Ok(logs::record_shared_log(log).await?)
}

#[tauri::command]
pub async fn logs_delete_shared_log(id: String, token: String) -> Result<()> {
    Ok(logs::delete_shared_log(id, token).await?)
}
